<launch>

    <arg name="delay_launch"          default="1.0" /> <!-- Delay of 1 second to let RViz come up -->
    <arg name="enable_thrust_control" default="false" />
    <arg name="sim_type"              default="rotors" />
 
    <arg name="world_name"            default="$(find caric_mission)/worlds/mbs.world" />
    <arg name="interest_points"       default="$(find caric_mission)/models/mbs/mbs_interest_points.pcd" />

    <arg name="enable_logging"        default="false" />
    <arg name="enable_ground_truth"   default="true" />
    <arg name="debug"                 default="false" />
    <arg name="gui"                   default="false" />
    <arg name="paused"                default="false" />
    <arg name="frame_id"              default="odom" />
    <arg name="verbose"               default="false" />
    <arg name="ppcom_hz"              default="50"/>
    <arg name="ppcom_file"            default="caric_ppcom_network_mbs.txt"/>

    <param name="/use_sim_time"       value="false"/>

    <!-- Launch the Gazebo World -->
    <env name="GAZEBO_MODEL_PATH"     value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models:$(find caric_mission)/models" />
    <env name="GAZEBO_RESOURCE_PATH"  value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models:$(find caric_mission)/models" />
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name"        value="$(arg world_name)" />
        <arg name="headless"          value="false"/>
        <arg name="debug"             value="$(arg debug)" />
        <arg name="paused"            value="$(arg paused)" />
        <arg name="gui"               value="$(arg gui)" />
        <arg name="verbose"           value="$(arg verbose)" />
    </include>

    <!-- Visualize MBS in RVIZ -->
    <param name="/model_path" type="string" value="$(find caric_mission)/models/mbs"/>
    <param name="/bounding_box_path" type="string" value="$(find caric_mission)/models/mbs/bounding_boxes"/>
    <node name="mesh_visualizer" pkg="caric_mission" type="mesh_visualizer.py" output="screen"
          launch-prefix="bash -c 'sleep $(arg delay_launch); $0 $@'">
    </node>

    <!-- Launch the MissionManager -->
    <node name="mission_manager" pkg="caric_mission" type="caric_mission_mission_manager_node"
          output="screen" launch-prefix="bash -c 'sleep delay_launch; echo LAUNCHING MISSION MANAGER...; $0 $@'">
          
          <param name="mission_duration" type="double" value="1800.0"/>
          <param name="log_dir"          type="string" value="$(find caric_mission)/logs/mbs/"/>
          <rosparam param="T_B_S">   [0.0, 0.0, 0.2, 1.0, 0.0, 0.0, 0.0]</rosparam>
          <rosparam param="T_S_L">   [0.0, 0.0, -0.03585, 1.0, 0.0, 0.0, 0.0]</rosparam>
          <param name="kf_knn_num"   type="double" value="5"/>
          <param name="kf_min_dis"   type="double" value="2.0"/>
          <param name="kf_min_ang"   type="double" value="10"/>
          <param name="kf_voxsize"   type="double" value="0.5"/>
    </node>

    <!-- Launch the ppcom router -->
    <node name="ppcom_router" pkg="caric_mission" type="ppcom_router.py" output="screen">
    </node>

    <!-- Spawning the ground station -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="tf_world_to_score_report" args="30 -40 0 0 0 0 world score_report"/>
    <include file="$(find caric_mission)/launch/spawn_gcs.launch">
        <arg name="namespace"         value="gcs" />
        <arg name="model"             value="$(find rotors_description)/urdf/gcs.xacro" />
        <arg name="ppcom_id"          value="gcs" />
        <arg name="ppcom_config"      value="$(find rotors_description)/ppcom_network/$(arg ppcom_file)" />
        <arg name="ppcom_hz"          value="$(arg ppcom_hz)" />
        <arg name="interest_points"   value="$(arg interest_points)" />
        <arg name="x"                 value="-2.0" />
        <arg name="y"                 value="08.0" />
        <arg name="z"                 value="20.0" />
    </include>

    <!-- Spawning jurong explorer drone -->
    <include file="$(find caric_mission)/launch/spawn_uav.launch">

        <arg name="mav_name"          value="firefly" />
        <arg name="namespace"         value="jurong" />
        <arg name="role"              value="explorer" />
        <arg name="model"             value="$(find rotors_description)/urdf/caric_explorer.xacro" />

        <arg name="ppcom_id"          value="jurong" />
        <arg name="ppcom_config"      value="$(find rotors_description)/ppcom_network/$(arg ppcom_file)" />
        <arg name="ppcom_hz"          value="$(arg ppcom_hz)" />

        <arg name="start_x"           value="0.0" />
        <arg name="start_y"           value="5.0" />

    </include>

    <node pkg="rviz" type="rviz" name="rviz_caric" args="-d $(find caric_mission)/rviz/caric_mbs.rviz" />

</launch>